File: go/tests/model_alignment_test.go
// model_alignment_test.go
//
// Verifies the CircleAI.ModelAlignment port (model_alignment.go):
//   - InMemoryAlignmentToolkit: Apply accepts reversible profiles and refuses
//     non-reversible ones; Revert removes an applied profile and reports the
//     not-applied / unknown-model cases; ListApplied returns a snapshot copy;
//     blank modelId/profileId are argument errors.
//   - RefuseAlignedPublishAuditor: refuses to publish once a profile is applied,
//     allows publish after it is reverted; nil toolkit is a ctor error.
//   - Null* fail-closed defaults refuse to apply/revert and always allow publish.

package circleai_test

import (
	"context"
	"strings"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func reversibleProfile(id string) circleai.AlignmentProfile {
	return circleai.AlignmentProfile{
		ProfileID:                id,
		Description:              "test",
		RefusalCategoriesRemoved: []string{"cat"},
		CreatedAtUTC:             time.Now().UTC(),
		IsReversible:             true,
	}
}

func TestInMemoryAlignmentToolkit_ApplyReversible(t *testing.T) {
	tk := circleai.NewInMemoryAlignmentToolkit()
	if tk.BackendID() != "in-memory" {
		t.Errorf("backend id: got %q", tk.BackendID())
	}
	ctx := context.Background()

	res, err := tk.Apply(ctx, "m1", reversibleProfile("p1"))
	if err != nil {
		t.Fatalf("apply err: %v", err)
	}
	if !res.Success || res.ProfileID != "p1" || res.FailureReason != "" {
		t.Errorf("apply result: %+v", res)
	}

	applied, err := tk.ListApplied(ctx, "m1")
	if err != nil {
		t.Fatalf("list err: %v", err)
	}
	if len(applied) != 1 || applied[0].ProfileID != "p1" {
		t.Errorf("list applied: %+v", applied)
	}
}

func TestInMemoryAlignmentToolkit_RefusesNonReversible(t *testing.T) {
	tk := circleai.NewInMemoryAlignmentToolkit()
	p := reversibleProfile("p1")
	p.IsReversible = false
	res, err := tk.Apply(context.Background(), "m1", p)
	if err != nil {
		t.Fatalf("apply err: %v", err)
	}
	if res.Success {
		t.Error("non-reversible must be refused")
	}
	if !strings.Contains(res.FailureReason, "Non-reversible") {
		t.Errorf("reason: %q", res.FailureReason)
	}
	// Nothing should have been recorded.
	applied, _ := tk.ListApplied(context.Background(), "m1")
	if len(applied) != 0 {
		t.Errorf("refused profile must not be stored, got %d", len(applied))
	}
}

func TestInMemoryAlignmentToolkit_Revert(t *testing.T) {
	tk := circleai.NewInMemoryAlignmentToolkit()
	ctx := context.Background()
	_, _ = tk.Apply(ctx, "m1", reversibleProfile("p1"))

	// Revert an applied profile.
	res, err := tk.Revert(ctx, "m1", "p1")
	if err != nil {
		t.Fatalf("revert err: %v", err)
	}
	if !res.Success {
		t.Errorf("revert should succeed: %+v", res)
	}
	if applied, _ := tk.ListApplied(ctx, "m1"); len(applied) != 0 {
		t.Errorf("profile should be gone, got %d", len(applied))
	}

	// Revert a profile not applied to a known model.
	_, _ = tk.Apply(ctx, "m1", reversibleProfile("p2"))
	res, _ = tk.Revert(ctx, "m1", "does-not-exist")
	if res.Success || !strings.Contains(res.FailureReason, "not applied") {
		t.Errorf("revert not-applied: %+v", res)
	}

	// Revert against an unknown model.
	res, _ = tk.Revert(ctx, "ghost", "p1")
	if res.Success || res.FailureReason != "Unknown model" {
		t.Errorf("revert unknown model: %+v", res)
	}
}

func TestInMemoryAlignmentToolkit_ArgErrors(t *testing.T) {
	tk := circleai.NewInMemoryAlignmentToolkit()
	ctx := context.Background()
	if _, err := tk.Apply(ctx, "   ", reversibleProfile("p1")); err == nil {
		t.Error("blank modelId should error on Apply")
	}
	if _, err := tk.Revert(ctx, "", "p1"); err == nil {
		t.Error("blank modelId should error on Revert")
	}
	if _, err := tk.Revert(ctx, "m1", ""); err == nil {
		t.Error("blank profileId should error on Revert")
	}
	if _, err := tk.ListApplied(ctx, ""); err == nil {
		t.Error("blank modelId should error on ListApplied")
	}
}

func TestInMemoryAlignmentToolkit_ListReturnsCopy(t *testing.T) {
	tk := circleai.NewInMemoryAlignmentToolkit()
	ctx := context.Background()
	_, _ = tk.Apply(ctx, "m1", reversibleProfile("p1"))
	got, _ := tk.ListApplied(ctx, "m1")
	got[0].ProfileID = "MUTATED"
	again, _ := tk.ListApplied(ctx, "m1")
	if again[0].ProfileID != "p1" {
		t.Error("ListApplied must return a snapshot copy, internal state was mutated")
	}
}

func TestRefuseAlignedPublishAuditor(t *testing.T) {
	tk := circleai.NewInMemoryAlignmentToolkit()
	aud, err := circleai.NewRefuseAlignedPublishAuditor(tk)
	if err != nil {
		t.Fatalf("ctor err: %v", err)
	}
	if aud.BackendID() != "refuse-aligned" {
		t.Errorf("backend id: got %q", aud.BackendID())
	}
	ctx := context.Background()

	// Clean model → publish ok.
	if err := aud.AssertOkToPublish(ctx, "m1"); err != nil {
		t.Errorf("clean model should publish: %v", err)
	}

	// After applying a profile → publish refused.
	_, _ = tk.Apply(ctx, "m1", reversibleProfile("p1"))
	err = aud.AssertOkToPublish(ctx, "m1")
	if err == nil {
		t.Fatal("aligned model must refuse publish")
	}
	if !strings.Contains(err.Error(), "alignment profile") || !strings.Contains(err.Error(), "m1") {
		t.Errorf("error should explain the refusal: %v", err)
	}

	// After reverting → publish ok again.
	_, _ = tk.Revert(ctx, "m1", "p1")
	if err := aud.AssertOkToPublish(ctx, "m1"); err != nil {
		t.Errorf("reverted model should publish: %v", err)
	}
}

func TestRefuseAlignedPublishAuditor_NilToolkit(t *testing.T) {
	if _, err := circleai.NewRefuseAlignedPublishAuditor(nil); err == nil {
		t.Error("nil toolkit should be a ctor error")
	}
}

func TestRefuseAlignedPublishAuditor_BlankModel(t *testing.T) {
	aud, _ := circleai.NewRefuseAlignedPublishAuditor(circleai.NewInMemoryAlignmentToolkit())
	if err := aud.AssertOkToPublish(context.Background(), ""); err == nil {
		t.Error("blank modelId should error")
	}
}

func TestModelAlignment_NullFailClosed(t *testing.T) {
	ctx := context.Background()
	tk := circleai.NullAlignmentToolkitInstance
	if tk.BackendID() != "null" {
		t.Errorf("null toolkit backend id: got %q", tk.BackendID())
	}

	res, _ := tk.Apply(ctx, "m1", reversibleProfile("p1"))
	if res.Success || res.ProfileID != "p1" || !strings.Contains(res.FailureReason, "no real backend") {
		t.Errorf("null apply: %+v", res)
	}
	res, _ = tk.Revert(ctx, "m1", "p9")
	if res.Success || res.ProfileID != "p9" || !strings.Contains(res.FailureReason, "nothing to revert") {
		t.Errorf("null revert: %+v", res)
	}
	if applied, _ := tk.ListApplied(ctx, "m1"); len(applied) != 0 {
		t.Errorf("null list should be empty, got %d", len(applied))
	}

	aud := circleai.NullAlignmentAuditorInstance
	if aud.BackendID() != "null" {
		t.Errorf("null auditor backend id: got %q", aud.BackendID())
	}
	if err := aud.AssertOkToPublish(ctx, "m1"); err != nil {
		t.Errorf("null auditor must always allow publish: %v", err)
	}
}

File: rust/tests/model_alignment_test.rs
//! model_alignment_test.rs
//!
//! Ports the behaviour of `CircleAI.ModelAlignment`
//! (`InMemoryModelAlignment.cs`, `Contracts.cs`, `NullImplementations.cs`):
//! reversible-only apply, revert semantics, the publish auditor that refuses
//! aligned models, argument validation, and the fail-closed `Null*` defaults.

use std::sync::Arc;

use chrono::Utc;
use circle_ai::model_alignment::{
    AlignmentError, AlignmentProfile, IAlignmentAuditor, IAlignmentToolkit, InMemoryAlignmentToolkit,
    NullAlignmentAuditor, NullAlignmentToolkit, RefuseAlignedPublishAuditor,
};

fn profile(id: &str, reversible: bool) -> AlignmentProfile {
    AlignmentProfile::new(
        id,
        "test profile",
        vec!["violence".to_string(), "self-harm".to_string()],
        Utc::now(),
        reversible,
    )
}

// ── InMemoryAlignmentToolkit ────────────────────────────────────────────────

#[test]
fn toolkit_backend_id() {
    assert_eq!(InMemoryAlignmentToolkit::new().backend_id(), "in-memory");
}

#[test]
fn apply_reversible_succeeds_and_lists() {
    let tk = InMemoryAlignmentToolkit::new();
    let r = tk.apply("model-a", &profile("p1", true)).unwrap();
    assert!(r.success);
    assert_eq!(r.profile_id, "p1");
    assert!(r.failure_reason.is_none());

    let applied = tk.list_applied("model-a").unwrap();
    assert_eq!(applied.len(), 1);
    assert_eq!(applied[0].profile_id, "p1");
}

#[test]
fn apply_non_reversible_is_refused() {
    let tk = InMemoryAlignmentToolkit::new();
    let r = tk.apply("model-a", &profile("p1", false)).unwrap();
    assert!(!r.success);
    assert_eq!(
        r.failure_reason.as_deref(),
        Some("Non-reversible alignment refused by InMemoryAlignmentToolkit")
    );
    // Nothing recorded.
    assert!(tk.list_applied("model-a").unwrap().is_empty());
}

#[test]
fn apply_empty_model_id_is_invalid_argument() {
    let tk = InMemoryAlignmentToolkit::new();
    let err = tk.apply("   ", &profile("p1", true)).unwrap_err();
    assert_eq!(err, AlignmentError::InvalidArgument("modelId required".into()));
}

#[test]
fn revert_removes_applied_profile() {
    let tk = InMemoryAlignmentToolkit::new();
    tk.apply("m", &profile("p1", true)).unwrap();
    tk.apply("m", &profile("p2", true)).unwrap();

    let r = tk.revert("m", "p1").unwrap();
    assert!(r.success);
    let applied = tk.list_applied("m").unwrap();
    assert_eq!(applied.len(), 1);
    assert_eq!(applied[0].profile_id, "p2");
}

#[test]
fn revert_unknown_model_reports_unknown() {
    let tk = InMemoryAlignmentToolkit::new();
    let r = tk.revert("ghost", "p1").unwrap();
    assert!(!r.success);
    assert_eq!(r.failure_reason.as_deref(), Some("Unknown model"));
}

#[test]
fn revert_not_applied_profile_reports_not_applied() {
    let tk = InMemoryAlignmentToolkit::new();
    tk.apply("m", &profile("p1", true)).unwrap();
    let r = tk.revert("m", "nope").unwrap();
    assert!(!r.success);
    assert_eq!(
        r.failure_reason.as_deref(),
        Some("Profile not applied to this model")
    );
}

#[test]
fn revert_requires_both_ids() {
    let tk = InMemoryAlignmentToolkit::new();
    assert_eq!(
        tk.revert("", "p1").unwrap_err(),
        AlignmentError::InvalidArgument("modelId required".into())
    );
    assert_eq!(
        tk.revert("m", "  ").unwrap_err(),
        AlignmentError::InvalidArgument("profileId required".into())
    );
}

#[test]
fn list_applied_unknown_model_is_empty() {
    let tk = InMemoryAlignmentToolkit::new();
    assert!(tk.list_applied("nobody").unwrap().is_empty());
}

#[test]
fn list_applied_empty_model_id_is_invalid_argument() {
    let tk = InMemoryAlignmentToolkit::new();
    assert_eq!(
        tk.list_applied(" ").unwrap_err(),
        AlignmentError::InvalidArgument("modelId required".into())
    );
}

// ── RefuseAlignedPublishAuditor ─────────────────────────────────────────────

#[test]
fn auditor_backend_id() {
    let tk = Arc::new(InMemoryAlignmentToolkit::new());
    assert_eq!(
        RefuseAlignedPublishAuditor::new(tk).backend_id(),
        "refuse-aligned"
    );
}

#[test]
fn publish_ok_when_no_profiles_applied() {
    let tk = Arc::new(InMemoryAlignmentToolkit::new());
    let auditor = RefuseAlignedPublishAuditor::new(tk);
    assert!(auditor.assert_ok_to_publish("clean-model").is_ok());
}

#[test]
fn publish_refused_when_profiles_applied() {
    let tk = Arc::new(InMemoryAlignmentToolkit::new());
    tk.apply("aligned-model", &profile("p1", true)).unwrap();
    let auditor = RefuseAlignedPublishAuditor::new(tk);

    let err = auditor.assert_ok_to_publish("aligned-model").unwrap_err();
    match err {
        AlignmentError::NotAllowed(msg) => {
            assert!(msg.contains("aligned-model"));
            assert!(msg.contains("1 alignment profile(s) applied"));
        }
        other => panic!("expected NotAllowed, got {other:?}"),
    }
}

#[test]
fn publish_ok_again_after_revert() {
    let tk = Arc::new(InMemoryAlignmentToolkit::new());
    tk.apply("m", &profile("p1", true)).unwrap();
    tk.revert("m", "p1").unwrap();
    let auditor = RefuseAlignedPublishAuditor::new(tk);
    assert!(auditor.assert_ok_to_publish("m").is_ok());
}

#[test]
fn auditor_rejects_empty_model_id() {
    let tk = Arc::new(InMemoryAlignmentToolkit::new());
    let auditor = RefuseAlignedPublishAuditor::new(tk);
    assert_eq!(
        auditor.assert_ok_to_publish("").unwrap_err(),
        AlignmentError::InvalidArgument("modelId required".into())
    );
}

// ── Null (fail-closed / fail-open) implementations ──────────────────────────

#[test]
fn null_toolkit_refuses_apply_and_revert() {
    let tk = NullAlignmentToolkit::INSTANCE;
    assert_eq!(tk.backend_id(), "null");

    let a = tk.apply("m", &profile("p1", true)).unwrap();
    assert!(!a.success);
    assert_eq!(
        a.failure_reason.as_deref(),
        Some("NullAlignmentToolkit: no real backend wired.")
    );
    assert_eq!(a.profile_id, "p1");

    let r = tk.revert("m", "p1").unwrap();
    assert!(!r.success);
    assert_eq!(
        r.failure_reason.as_deref(),
        Some("NullAlignmentToolkit: nothing to revert.")
    );

    assert!(tk.list_applied("m").unwrap().is_empty());
}

#[test]
fn null_auditor_always_allows_publish() {
    let auditor = NullAlignmentAuditor::INSTANCE;
    assert_eq!(auditor.backend_id(), "null");
    assert!(auditor.assert_ok_to_publish("anything").is_ok());
}
